Implement group types feature

bitterdev opened the pull request.

bitterdev commented:
I have added another feature into this big PR (see https://github.com/concrete5/concrete5/pull/9373/commits/3c1c36f83cdf4b4c015ff65ca85e1d423cb75075)

This feature allows theme developers to provide multiple content swap files.

aembler commented:
This is very exciting... can you run the tests locally and figure out why the groups/users tests are failing? It's not a surprise that they're failing, there's a lot of new code here. If we have to modify the tests a little bit that's ok, but we should try and ensure that this is as backward compatible as possible.

aembler commented:
Additionally when I try to test this locally it complains about `fetchAllAssociative` not being a valid method. Is this something new in doctrine? Even after running composer update it's not present.

bitterdev commented:
> Additionally when I try to test this locally it complains about `fetchAllAssociative` not being a valid method. Is this something new in doctrine? Even after running composer update it's not present.

https://www.doctrine-project.org/2020/11/17/dbal-3.0.0.html

fetchAll is depreceated in DBAL 3 and since DBAL 2.12 tagged with the @deprecated doc-block. I'm wondering why you don't have this methods even after running composer update because concreteCMS's composer.json is requesting DBAL 2.5 and this is part of it. 
I have checked out concreteCMS repo from scratch and run composer install and i have all this methods available. 
I really don't have any idea but when it's a problem i can downgrade it to fetchAll.

bitterdev commented:
![Bildschirmfoto 2021-03-01 um 09 41 22](https://user-images.githubusercontent.com/15018219/109472359-4fe1f180-7a72-11eb-8f85-8536a29054e0.png)

Why i run the tests locally i got other issues from files that i don't have affected but no issue that are related to groups.

aembler commented:
Check the tests that run in Travis. For some reason your tests aren't completing

There are group and user related failures. You may have to force phpunit to just run the failing tests by using the --filter option

aembler commented:
Are you sure you haven't upgraded to doctrine 3.0 in your local development environment ?

bitterdev commented:
Hi Andrew, i have replaced all not supported PDO calls. It seems that i had a newer PDO version installed on my dev environment. I think the reason for that was that i have used composer 2. As i have downgraded composer and updated the composer dependencies of concrete all depreceation tags are away.

bitterdev commented:
Added a core fix (user selector had no search index). Will now fix the auto tests.

aembler commented:
I think some of the dependencies here got downgraded (see the copyright dates in CKEditor) but that's minor and I can fix that.

This is looking really good. There's a lot here obviously so expect to see some tickets around some changes in the future, but nothing that needs to hold this up.
